### DISCUSSION TITLE
coloring a map

### DISCUSSION THREAD

**Discussion by fkv1** [networkx/networkx#6672] [Algorithms]
I needed to color regions in a geographic map and remembered the 4 color theorem. Given the relatively big number (534) of regions in my map, I didn't want to do it by hand and searched for an existing implementation in python. I came across implementations of some algorithms in or using the networkx module. I tried all of them and every result had a fifth color in it. I ended up developing my own algorithm, using a brute force method similar to the "greedy" algorithm, but with some logic to avoid a fifth color while keeping the running time acceptable. Coloring the 534 nodes (regions) in my graph (map) takes around 1½ minutes on my computer. That's fine for me because I need to color my map only once.

Assuming that my implementation might be useful for others as well, I transferred it into a networkx function, but uploading the files to your project isn't trivial. In most bugtrackers I came across in my life, I could write a feature request and attach the files. On Github, I need to create a fork and whatnot. I gave up when github.com rejected my RSA key for an unknown reason. Neither can I attach the files to this message, because .py files aren't among the "supported file types".

Let me know if you are interested in my implementation, and if so, how I can contribute it. Of course, if you come up with a faster algorithm and implement it yourself, that's also fine.

**Comment by dschult** [comment]
Depending on how big your code is you could add it to a comment in this thread using copy/paste and enclosing the python code inside triple backquotes -- adding the keyword "python" formats it all nicely:

```
```python
def myfunct(x, y):
    return x + y
```  # this triple backspace indicates the end of the formatted input
```

shows up as:
```python
def myfunct(x, y):
    return x + y
```

If your code is quite long, you can include a link in the comment to an online spot that has that code. I also believe that a `patch`-file (also called a diff file) can be attached to a comment -- but I haven't tried that.

**Comment by fkv1** [comment]
Thanks.

networkx/algorithms/coloring/brute_force_coloring.py:

```python
"""
Brute-force vertex coloring using compact-border strategy
"""

import networkx as nx

__all__ = ["brute_force_color"]


def brute_force_color(G, maxcolors=4, balance=False, startnode=None):
    """A brute-force coloring algorithm designed for 4-coloring real (e.g.
    geographic) maps using the strategy a human would normally use, i.e. to
    start on one side and keep the border between already colored and not yet
    colored nodes (countries) compact. This prevents deep recursion into
    impossible combinations and thus keeps performance acceptable for most maps.

    Parameters
    ----------
    G: NetworkX graph

    maxcolors: non-negative integer
         The maximum number of colors. If 0 or None, the smallest possible
         number of colors will be used. This takes slightly longer than when
         it is supplied.

    balance: boolean
         Whether to try to balance the number of nodes for each color. If
         False, there is a bias towards lower-indexed colors.

    startnode: NetworkX node
         Where to start calculation. If not supplied, a feasible start node is
         calculated. This node receives color 0.

    Returns
    -------
    If successful, a dictionary with keys representing nodes and values representing
    colors, starting with 0. Otherwise, None is returned.

    Raises
    ------
    NetworkXError
        If startnode is not a node in the graph.
    """

    if startnode is not None:
        if startnode not in G:
            raise nx.NetworkXError("startnode is not in the graph.")
    if not len(G):
        return {}
    n_in_order = []
    startind = 0
    while len(n_in_order) < len(G): # loop over unconnected parts of the graph
        if startnode is None:
            mincnt = None
            for n in G:
                if n not in n_in_order:
                    cnt = -2 * len(G.adj[n])
                    for n2 in G.adj[n]:
                        cnt += len(G.adj[n])
                    if mincnt is None or cnt < mincnt:
                        mincnt = cnt
                        startnode = n

        # determine sequence of nodes to calculate
        n_in_order.append(startnode)
        nodes_in_part = 1
        while len(n_in_order) < len(G):
            # determine stripe bordering already colored area
            stripe = set()
            for n in n_in_order[startind:]:
                for nbr in G.adj[n]:
                    if nbr not in n_in_order[startind:]:
                        stripe.add(nbr)
            if not len(stripe):
                break # path containes another unconnected part, handled in outer loop
            # calculate score for each node in stripe
            minscore = None
            optim_n = None
            for n in stripe:
                score = 0
                for n2 in n_in_order[startind:] + [n]:
                    on_border = 0
                    for nbr in G.adj[n2]:
                        if nbr in n_in_order[startind:]:
                            score -= 5
                        else:
                            on_border = 11
                            score += 1 # one border segment (edge crossing border); the more of them, the worse
                    score += on_border # the more nodes along border, the worse
                if minscore is None or score < minscore:
                    minscore = score
                    optim_n = n
            n_in_order.append(optim_n)
            nodes_in_part += 1
        startnode = None
        startind += nodes_in_part

    for numcolors in (maxcolors,) if maxcolors else (1,2,3,4):
        # greedy-color nodes in sequence and step back whenever getting stuck
        colors = dict((n, None) for n in G)
        color_transl = [[r for r in range(numcolors)]] * len(n_in_order)
        i = 0
        while i < len(n_in_order):
            n = n_in_order[i]
            transl_ind = -1 if colors[n] is None else color_transl[i].index(colors[n])
            if transl_ind + 1 == numcolors:
                colors[n] = None
                if i == numcolors:
                    break
                i -= 1
                continue
            colors[n] = color_transl[i][transl_ind + 1]
            color_available = True
            for nbr in G.adj[n]:
                if colors[nbr] == colors[n]:
                    color_available = False
                    break
            if color_available:
                i += 1
                if balance and i < len(n_in_order):
                    colcnt = [0] * numcolors
                    for j in range(i):
                        colcnt[colors[n_in_order[j]]] += 1
                    color_transl[i] = sorted([c for c in range(numcolors)], key=lambda c: colcnt[c])
        if colors[n_in_order[-1]] is not None:
            return colors
    return None
```

```diff -u networkx/algorithms/coloring/__init__.py.orig networkx/algorithms/coloring/__init__.py
--- networkx/algorithms/coloring/__init__.py.orig       2023-04-29 22:22:33.510194000 +0200
+++ networkx/algorithms/coloring/__init__.py    2023-04-29 22:34:54.627057000 +0200
@@ -1,4 +1,5 @@
 from networkx.algorithms.coloring.greedy_coloring import *
 from networkx.algorithms.coloring.equitable_coloring import equitable_color
+from networkx.algorithms.coloring.brute_force_coloring import *
 
-__all__ = ["greedy_color", "equitable_color"]
+__all__ = ["greedy_color", "equitable_color", "brute_force_color"]
```

PS: edited because startnode may be 0 (zero).
PS2: edited again on Jul 3 to fix a bug that occurs when the number of colors needed equals the number of nodes

**Comment by fkv1** [comment]
Are there any plans to incorporate this in your project? Or do you think it would fit better into another project? Of course, I could publish a separate module just for this, but that would feel clumsy.

**Reply by dschult** [reply]
I'm interested, but there is a lot to do. 
Could you describe how this connects to the map coloring problem you describe in the doc_string? As written it looks like G can be any graph, and that surely can't be true. 

Do you have any examples of usage on a simple small problem? (we could include those in the doc_strings)

Do you have any tests where the solution is known? Something to make sure this is doing what you think it is doing? (so when it gets updated later we can make sure it still works?

Sorry for the late reply (submitted right at the end of the semester) and thanks for the nudge.

**Comment by fkv1** [comment]
> Could you describe how this connects to the map coloring problem you describe in the doc_string?

It solves it. The algorithm is certainly not the fastest theoretically possible, but I have not seen any implementation of a faster algorithm yet, let alone with a balancing option.

> As written it looks like G can be any graph, and that surely can't be true.

I know little about graph theory. I have no idea what weird kinds of graphs exist (like multi-dimensional or with duplicate edges). It should work on all graphs that represent geographic maps, with one vertex for each region, and every edge representing that the connected regions share at least one border with length > 0.

With disjunct regions (with exclaves such as Russia's Kaliningrad oblast), 4 colors may not be enough, but that's not the algorithm's fault.

> Do you have any examples of usage on a simple small problem? (we could include those in the doc_strings)

The simplest example of a map that requires 4 colors is a country surrounded by 3 other countries, so that each country neighbors each other.  The graph representation is 4 vertices and 6 edges:

```python
countries = [
    { 'name' : 'Fribourg',
      'neighbors': ['Bern', 'Vaud', 'Neuchâtel']
    },
    { 'name' : 'Bern',
      'neighbors': ['Fribourg', 'Vaud', 'Neuchâtel']
    },
    { 'name' : 'Vaud',
      'neighbors': ['Fribourg', 'Bern', 'Neuchâtel']
    },
    { 'name' : 'Neuchâtel',
      'neighbors': ['Fribourg', 'Bern', 'Vaud']
    }
]

colors = ['red', 'green', 'blue', 'yellow']

import networkx as nx
G = nx.Graph()
for c in countries:
    G.add_node(c['name'])
for c in countries:
    for n in c['neighbors']:
        G.add_edge(c['name'], n)

dict = nx.brute_force_color(G)
if dict is None:
    print('failed')
else:
    for c in dict:
        print (c + ': ' + colors[dict[c]])
```

output:
```
Fribourg: red
Bern: green
Vaud: yellow
Neuchâtel: blue
```

Funnily enough, this minimal example revealed a bug that doesn't occur with more complicated examples. Meanwhile I've edited the bugfix into the source code in my previous post.

> Do you have any tests where the solution is known? Something to make sure this is doing what you think it is doing? (so when it gets updated later we can make sure it still works?

I hope that whoever wants to change the function thinks twice and knows what they do. It has several parameters, and for every combination of parameters there are various edge cases to consider, so it's essentially impossible to cover them all with regression tests. Rare edge cases are those where bugs are most probably introduced and where they remain undetected for a long time.

While a passed regression test does not imply all is fine, the opposite is also not true. All graphs with edges have multiple solutions. E.g. you can swap all reds and greens and the solution would still be correct, but the test woulds say it's broken.

I guess the performance (runtime) is what future developers will most likely try to improve, and by trying that they can even make it worse. This is impossible to cover with regression tests, because the runtime depends on the hardware etc.

> Sorry for the late reply (submitted right at the end of the semester) and thanks for the nudge.

I was only worried because old forum posts tend to fall into oblivion, and even some "proper" pull requests have been open for 10 years.

**Comment by fkv1** [comment]
7 months later, I feel like it's time for another reminder. I'm pretty sure that there are people who can use this (coloring a map isn't such an exotic task), and I have no idea where else to publish my algorithm so that they can find it. And an algorithm alone won't do the job, they need an implementation. It's already here, just waiting to be committed. Is this an insurmountable obstacle?

**Comment by fkv1** [comment]
For those who aren't sure what I'm talking about, here's the map for which I needed to come up with the algorithm:
combine "bio.volki." and "at/fund.py?KARTE=KN" as the URL
(I don't want Google to index that web application, because it's not yet finished.)
